Polyadic.Json/NewType/JsonNewTypeJsonConverterFactory.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Polyadic.Json.NewType
+{
+    internal sealed class JsonNewTypeJsonConverterFactory : JsonConverterFactory
+    {
+        public override bool CanConvert(Type typeToConvert) => Attribute.IsDefined(typeToConvert, typeof(JsonNewTypeAttribute));
+
+        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+        {
+            var metadata = GetMetadata(typeToConvert, options);
+            return (JsonConverter)Activator.CreateInstance(
+                typeof(JsonNewTypeConverter<,>).MakeGenericType(typeToConvert, metadata.InnerType),
+                metadata)!;
+        }
+
+        private static NewTypeMetadata GetMetadata(Type typeToConvert, JsonSerializerOptions options)
+        {
+            var constructor = FindConstructor(typeToConvert);
+            var innerType = constructor.GetParameters()[0].ParameterType;
+            var converter = options.GetConverter(innerType);
+            var valueProperty = FindValueProperty(typeToConvert, innerType);
+            return new NewTypeMetadata(innerType, converter, constructor, valueProperty);
+        }
+
+        private static ConstructorInfo FindConstructor(Type typeToConvert)
+        {
+            var constructorCandidates = typeToConvert.GetConstructors().Where(IsUnary).ToImmutableArray();
```

Comment:
Would you like to extract these into functions?

Reply:
759fc7d98f494d05daa1cc5e7e6c891b47d7527d